### PR TITLE
feat(plugin-pnpm): Content-addressable storage support

### PR DESCRIPTION
**Pull request by larixer**
**What's the problem this PR addresses?**

This PR implements content-addressable storage support for `pnpm` linker via `nmMode: hardlinks-global` config setting.

Integrates PR: #4532 

**How did you fix it?**

The PR is reusing `hardlinks-global` implementation from `node-modules` linker.

`pnpm` linker link step timings on one of my rather big projects:
1. From `master` (no hardlinks store): 18.9 secs
2. From this PR (no hardlinks store): 14.8 secs
3. From this PR (`nmMode: hardlinks-global`, empty store): 37.8 secs
4. From this PR (`nmMode: hardlinks-global`, filled store): 15.2 secs, it is a bit slower than wihout hardlinks, because content addressable store needs more fs operations on packages without checksums, which is the case for conditional packages like `fsevents`

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-nm/sources/NodeModulesLinker.ts [hunk 1]
```diff
@@ -756,14 +770,15 @@ type DirEntry = {
...
+  mtimeMs?: number;
 } | {
   kind: DirEntryKind. DIRECTORY;
 } | {
   kind: DirEntryKind.SYMLINK;
   symlinkTo: PortablePath;
 };
 
-const copyPromise = async (dstDir: PortablePath, srcDir: PortablePath, {baseFs, globalHardlinksStore, nmMode, packageChecksum}: {baseFs: FakeFS<PortablePath>, globalHardlinksStore: PortablePath | null, nmMode: {value: NodeModulesMode}, packageChecksum: string | null}) => {
+export const copyPromise = async (dstDir: PortablePath, srcDir: PortablePath, {baseFs, hardlinksStorePath, nmMode, packageChecksum}: {baseFs: FakeFS<PortablePath>, hardlinksStorePath: PortablePath | null, nmMode: {value: NodeModulesMode}, packageChecksum: string | null}) => {
```
Comment (arcanis) [inline packages/plugin-nm/sources/NodeModulesLinker.ts:781]: I'm really not fan of having a separate `copy` function... It was fine when it was bound to the nm linker, but if it leaks into the others it starts to become concerning imo 🤔

**Review by arcanis — COMMENTED** [review]
I want to take some time to think about this; I may experiment myself with implementing some logic into the main copyPromise (perhaps it won't work anyway, but I'd like to better understand why).

**Comment by TAUnionOtto** [comment]
sorry but……any update here?

**Comment by larixer** [comment]
Closed in favour of: #4586, although the latter has worse performance

**Pull request closed without merging**